File: src/main/java/com/jacuum/leaderboard/JsonLeaderboard.java
package com.jacuum.leaderboard;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

public final class JsonLeaderboard implements Leaderboard {

    private final Path file;
    private final ObjectMapper mapper;
    private final TypeReference<List<LeaderboardEntry>> type;

    public JsonLeaderboard(final Path file) {
        this.file = file;
        this.mapper = new ObjectMapper()
            .enable(SerializationFeature.INDENT_OUTPUT);
        this.type = new TypeReference<>() {};
    }

    @Override
    public synchronized List<LeaderboardEntry> entries() throws Exception {
        if (!Files.exists(this.file)) return List.of();
        return this.mapper.readValue(this.file.toFile(), this.type);
    }

    @Override
    public synchronized void save(final LeaderboardEntry entry) throws Exception {
        final List<LeaderboardEntry> current = new ArrayList<>(entries());
        current.add(entry);
        this.mapper.writeValue(this.file.toFile(), current);
    }
}

File: src/main/kotlin/com/jacuum/leaderboard/JsonLeaderboard.kt
package com.jacuum.leaderboard

import com.fasterxml.jackson.databind.SerializationFeature
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import java.nio.file.Files
import java.nio.file.Path

class JsonLeaderboard(private val file: Path) : Leaderboard {

    private val mapper = jacksonObjectMapper()
        .enable(SerializationFeature.INDENT_OUTPUT)

    @Synchronized
    override fun entries(): List<LeaderboardEntry> {
        if (!Files.exists(file)) return emptyList()
        return mapper.readValue(file.toFile())
    }

    @Synchronized
    override fun save(entry: LeaderboardEntry) {
        val current = entries().toMutableList()
        current.add(entry)
        mapper.writeValue(file.toFile(), current)
    }
}
